README.md
![pipeline](https://gitlab.uib.gmbh/uib/python-opsi/badges/v4.2/pipeline.svg)
![coverage](https://gitlab.uib.gmbh/uib/python-opsi/badges/v4.2/coverage.svg)

# python-opsi
This is the Python library behind the client management-tool [opsi](http://www.opsi.org/).


## License
This library is released under the AGPLv3 and the copyright belongs to
uib GmbH if this is not noted otherwise in the file itself.


## Documentation
You can use [Sphinx](http://sphinx-doc.org/) to build the documentation.
If you are looking for information on how to setup or configure an opsi
system please get the _getting started_ from opsi.org.

### Building the documentation
First we create the API documentation from the Python files:

```bash
sphinx-apidoc --separate --output-dir=doc/src OPSI/
```

After that we can build the documentation:

```bash
sphinx-build -b html -d doc/_build/doctrees doc/src/ doc/python-opsi/
```

After that you will find the documentation in the folder `doc/python-opsi`.


## Requirements
Opsi relies on a mix of Python libraries and system tools that need to
be installed.

The dependencies can be found in `pyproject.toml`.
Please use pip or your distributions recommended tool for the installation of
these.


### Installing on Ubuntu
Installing the depedencies via apt-get:

```bash
apt-get install python3-dev python3-twisted python3-magic python3-pycryptodome python3-newt python3-pampy python3-openssl python3-mysqldb python3-sqlalchemy python3-psutil iproute2 lshw mysql-server libmysqlclient-dev
python3.7 -m pip install mysqlclient
```

## Packaging
You need `python poetry` to build sdist / wheel / Debian and RPM packages.

Build sdist and wheel package:

```bash
poetry install
poetry build
```

Build debian package:

```bash
apt install dpkg-dev
poetry install
poetry run opsi-dev-tool --deb-create-pkg .
```

## Testing
Tests can be found in the `tests` folder. We use [pytest](http://pytest.org/) for our tests.


### Configuring database for test
Testing the MySQL backend requires a license file for most of the tests.

To run tests with MySQL as a backend you need to install and configure
your MySQL server first.
You then need to create a user and database for the tests.
Please follow the corresponding guides of your distribution and/or MySQL
to do so.

**Warning:** The tests will drop every table on the configured database
so make sure you are not running things against your production database!

It is possible to let opsi create a database for you by running `opsi-setup --configure-mysql` and then re-use the configuration from `/etc/opsi/backends/mysql.conf`.

To configure the tests copy the example configuration to `tests/testOPSI/Backends/config.py`:

```bash
cp tests/Backends/config.py.example tests/Backends/config.py
```

In this file fill the dict `MySQLconfiguration` with the settings for your test database.
If your are reusing the values from `/etc/opsi/backends/mysql.conf` you can copy the content of `config` to it.


### Run tests
Tests can then be run with:

```bash
poetry install
poetry run pytest
```

### Running Tests on local machine with docker
You need docker, git, python3 and python3-pip installed.

First install poetry:

```bash
pip3 install poetry
```

To run all tests you need a modules file under /etc/opsi of the machine (set the rights for your user, that will run the tests).

You will find a file under `tests/testOPSI/Backends/config.py.gitlabci` copy this file as `config.py` in the same directory.

Start a docker container with mysql for tests:

```bash
docker run --detach --name=mysql --env="MYSQL_ROOT_PASSWORD=opsi" --env="MYSQL_DATABASE=opsi" mysql:latest
```

Grab the ip of your new container with:

```bash
docker inspect mysql
```

and patch your `tests/testOPSI/Backends/config.py` with the new host information for mysql.

Disable strict mode from mysql:

```bash
mysql --host=172.17.0.2 --user=root --password=opsi -e "SET GLOBAL sql_mode = 'NO_ENGINE_SUBSTITUTION';"
```

Change host ip from that what you have seen in docker inspect of your machine.

If you want to run your tests under Ubuntu 18.04 you need also a pip update from ppa:

```bash
apt -y install software-properties-common
add-apt-repository ppa:ci-train-ppa-service/3690
apt -y install python-pip=9.0.1-2.3~ubuntu1.18.04.2~ubuntu18.04.1~ppa202002141134
```

Last step for running:

You need the files from opsi-server for the tests. If you have also cloned opsi-server in the same directory like python-opsi you can set a symbolic link to the data-files:

```bash
ln -s ../opsi-server/opsi-server_data/etc data
```

Now you can install your venv over poetry:

```bash
poetry install
```

Now run the tests:

```bash
poetry run pytest
```

## Contributing
Contributions are welcome.

If you find any security problem please inform us (info@uib.de) before disclosing the security vulnerability in public.


### Translation
Translations are made via [Transifex](https://www.transifex.com/opsi-org/opsiorg/) and the corresponding resource is located [here](https://www.transifex.com/opsi-org/opsiorg/python-opsi/).


### Tests
Please provide tests or a guide on how to test with your contributions.
After applying your code changes all tests must pass.


### Coding Style
Please use conventions described in [PEP 8 -- Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/).
Deviating from this, indentation has to be done with hard tabs.

For general information about webservice methods please refer to the [manual](http://download.uib.de/opsi4.0/doc/html/en/opsi-manual/opsi-manual.html#opsi-manual-api-datastructure-opsi).


#### Semi-Automated Quality Checks
There is a script that runs `pylint`, `flake8` and all the tests.

```bash
poetry install
poetry run ./run_qa.sh
```

The script will not display any problems reported by `pylint` or
`pep8` but instead creates the files `pylint.txt` and `pep8.txt`.
You then can check the corresponding output.

It will also run all tests and create a coverage from those tests as
`coverage.xml`.


### Documentation
Documentation should be provided for any non-intuitive or complex part.
Please provide the documentation either directly as Python docstrings or
provide it in the form of documents inside the `doc` folder.
The documentation should be integrated into the documentation that is
built with Sphinx.


debian/changelog
python-opsi (4.1.1.103-1) stable; urgency=medium

  * WindowsDrivers.py: changed logging to be more like in show_drivers.py

 -- Mathias Radtke <m.radtke@uib.de> Thu, 25 Mar 2021 16:22:00 +0100

python-opsi (4.1.1.102-1) stable; urgency=medium

  * WindowsDrivers.py: added more logging on searched byAudit vendor and model directories

 -- Mathias Radtke <m.radtke@uib.de> Thu, 25 Mar 2021 10:00:00 +0100

python-opsi (4.1.1.101-1) stable; urgency=medium

  * updated hwaudit locales to include TPM data

 -- Nils Doerrer <n.doerrer@uib.de>  Fri, 19 Mar 2021 09:23:33 +0100

python-opsi (4.1.1.100-1) stable; urgency=medium

  * introduced TPM section and fields for MacOS data in opsihwaudit.conf

 -- Nils Doerrer <n.doerrer@uib.de>  Fri, 12 Mar 2021 09:29:43 +0100

python-opsi (4.1.1.99-1) testing; urgency=medium

  * adding more opsi-bootimage.append parameters

 -- Mathias Radtke <m.radtke@uib.de>  Wed, 27 Jan 2021 17:08:46 +0100

python-opsi (4.1.1.98-1) testing; urgency=medium

  * Posix.py: not using -D option in updated sfdisk anymore 

 -- Mathias Radtke <m.radtke@uib.de>  Wed, 27 Jan 2021 16:11:55 +0100

python-opsi (4.1.1.97-1) testing; urgency=medium

  * Validation of filters and attributes in backend getObject methods to prevent SQL injection attacks
  * OPSI.System.Posix: removed sfdisk downgrade functions

 -- Mathias Radtke <m.radtke@uib.de>  Mon, 14 Dec 2020 12:22:19 +0100

python-opsi (4.1.1.96-1) stable; urgency=medium

  * Fixing sql injection issue in backend

 -- Jan Schneider <j.schneider@uib.de>  Tue, 01 Dec 2020 15:34:58 +0100

python-opsi (4.1.1.95-1) stable; urgency=medium

  * setGeneralConfig: Raise Exception if given objectId is not found in backend
  * JSONRPCBackend: Small fix in creating instancemethods.

 -- Erol Ueluekmen <e.ueluekmen@uib.de>  Mon, 07 Sep 2020 09:06:00 +0200

python-opsi (4.1.1.94-1) stable; urgency=medium

  * mysql strict mode support

 -- Erol Ueluekmen <e.ueluekmen@uib.de>  Mon, 13 Jul 2020 01:45:02 +0200

python-opsi (4.1.1.93-1) stable; urgency=medium

  * autosetupexcludes could be empty.
  * ProductRepositoryInfo refactoring.

 -- Erol Ueluekmen <e.ueluekmen@uib.de>  Mon, 29 Jun 2020 22:51:52 +0200

python-opsi (4.1.1.92-1) stable; urgency=medium

  * small typofix in Updater.py

 -- Erol Ueluekmen <e.ueluekmen@uib.de>  Mon, 29 Jun 2020 16:59:53 +0200

python-opsi (4.1.1.91-1) stable; urgency=medium

  * Follow links at PackageContentFile generation after opsi-package installation

 -- Erol Ueluekmen <e.ueluekmen@uib.de>  Fri, 26 Jun 2020 16:08:05 +0200

python-opsi (4.1.1.90-1) stable; urgency=medium

  * PackageUpdater: autosetup excludes implemented.

 -- Erol Ueluekmen <e.ueluekmen@uib.de>  Fri, 22 May 2020 16:11:12 +0200

python-opsi (4.1.1.89-1) testing; urgency=medium

  * OpsiPXEConfd: setting eliloMode to x64 when shim is present in
    clientconfig.dhcpd.filename

 -- Mathias Radtke <m.radtke@uib.de>  Mon, 06 Apr 2020 18:06:48 +0200

python-opsi (4.1.1.88-1) stable; urgency=medium

  * version bumped to cleanup rpm builds in buildservice.

 -- Erol Ueluekmen <e.ueluekmen@uib.de>  Wed, 05 Feb 2020 11:02:12 +0100

python-opsi (4.1.1.87-1) stable; urgency=medium

  * DHCPD-Parser small fix for multiple domain-search entries
  * small fix for centos/rhel 8 build

 -- Erol Ueluekmen <e.ueluekmen@uib.de>  Tue, 04 Feb 2020 00:56:00 +0100

python-opsi (4.1.1.86-1) testing; urgency=medium

  * OPSI.Util.Task.UpdatePackages.Updater: Fix throwing an exception when
    mismatching md5sums are detected.
  * OPSI.Util.Task.UpdatePackages.Updater: Error on wrong md5sum in remote
    repository.

 -- Niko Wenselowski <n.wenselowski@uib.de>  Tue, 03 Dec 2019 11:31:01 +0100

python-opsi (4.1.1.85-1) testing; urgency=medium

  * OPSI.Util.Task.UpdatePackages.Updater: throwing exception when md5sum
    missmatched an no zsync is available

 -- Mathias Radtke <m.radtke@uib.de>  Wed, 20 Nov 2019 13:37:21 +0100

python-opsi (4.1.1.84-1) testing; urgency=medium

  * OPSI.Util.File.InfFile: Small refactoring in device type detection.
  * OPSI.System.Posix: Support for using the 'ip' command to parse network
    information. By default 'ifconfig' is tried but in case of absence 'ip'
    will be used.
    This requires iproute2 4.15 or newer.

 -- Niko Wenselowski <n.wenselowski@uib.de>  Thu, 14 Nov 2019 13:13:56 +0100

python-opsi (4.1.1.83-1) testing; urgency=medium

  * Fix regular expression pattern of HD audio device drivers.
    This has been broken since 4.1.1.63-1.

 -- Niko Wenselowski <n.wenselowski@uib.de>  Mon, 04 Nov 2019 14:51:34 +0100

python-opsi (4.1.1.82-1) testing; urgency=medium

  * 30_kiosk.conf: Getting the product groups now works with broken configs.
  * MySQL backend: Changed SOFTWARE_CONFIG.config_id to bigint to allow for
    more entries. An update routine has been added that can be executed by
    running opsi-setup --update-mysql.

 -- Niko Wenselowski <n.wenselowski@uib.de>  Thu, 17 Oct 2019 13:19:31 +0200

python-opsi (4.1.1.81-1) stable; urgency=medium

  * Use an User-Agent setting in accordance to RFC 2616 in more places.

 -- Niko Wenselowski <n.wenselowski@uib.de>  Fri, 11 Oct 2019 15:22:25 +0200

python-opsi (4.1.1.80-1) stable; urgency=medium

  * OPSI.System.Posix.getActiveSessionIds: Support parsing output where the

OPSI/Types.py
# -*- coding: utf-8 -*-

# This file is part of python-opsi.
# Copyright (C) 2006-2019 uib GmbH <info@uib.de>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
Type forcing features.

This module contains various methods to ensure force a special type
on an object.

:copyright:	uib GmbH <info@uib.de>
:license: GNU Affero General Public License version 3
"""

import datetime
import ipaddress
import os
import re
import sys
import time
import types

from OPSI.Logger import Logger

__all__ = (
	'args', 'forceActionProgress', 'forceActionRequest',
	'forceActionRequestList', 'forceActionResult', 'forceArchitecture',
	'forceArchitectureList', 'forceAuditState', 'forceBool', 'forceBoolList',
	'forceConfigId', 'forceDict', 'forceDictList', 'forceDomain',
	'forceEmailAddress', 'forceFilename', 'forceFloat', 'forceFqdn',
	'forceGroupId', 'forceGroupIdList', 'forceGroupType', 'forceGroupTypeList',
	'forceHardwareAddress', 'forceHardwareDeviceId', 'forceHardwareVendorId',
	'forceHostAddress', 'forceHostId', 'forceHostIdList', 'forceHostname',
	'forceIPAddress', 'forceInstallationStatus', 'forceInt', 'forceIntList',
	'forceIpAddress', 'forceLanguageCode', 'forceLanguageCodeList',
	'forceLicenseContractId', 'forceLicenseContractIdList', 'forceLicensePoolId',
	'forceLicensePoolIdList', 'forceList', 'forceNetmask',
	'forceNetworkAddress', 'forceObjectClass', 'forceObjectClassList',
	'forceObjectId', 'forceObjectIdList', 'forceOct', 'forceOpsiHostKey',
	'forceOpsiTimestamp', 'forcePackageCustomName', 'forcePackageVersion',
	'forcePackageVersionList', 'forceProductId', 'forceProductIdList',
	'forceProductPriority', 'forceProductPropertyId',
	'forceProductPropertyType', 'forceProductTargetConfiguration',
	'forceProductType', 'forceProductVersion', 'forceProductVersionList',
	'forceRequirementType', 'forceSoftwareLicenseId',
	'forceSoftwareLicenseIdList', 'forceTime', 'forceUnicode',
	'forceUnicodeList', 'forceUnicodeLower', 'forceUnicodeLowerList',
	'forceUnicodeUpper', 'forceUniqueList', 'forceUnsignedInt', 'forceUrl'
)

encoding = sys.getfilesystemencoding()
logger = Logger()

_HARDWARE_ID_REGEX = re.compile(r'^[a-fA-F0-9]{4}$')
_OPSI_TIMESTAMP_REGEX = re.compile(r'^(\d{4})-?(\d{2})-?(\d{2})\s?(\d{2}):?(\d{2}):?(\d{2})\.?\d*$')
_OPSI_DATE_REGEX = re.compile(r'^(\d{4})-?(\d{2})-?(\d{2})$')
_FQDN_REGEX = re.compile(r'^[a-z0-9][a-z0-9\-]{,63}\.((\w+\-+)|(\w+\.))*\w{1,63}\.\w{2,16}\.?$')
_HARDWARE_ADDRESS_REGEX = re.compile(r'^([0-9a-f]{2})[:-]?([0-9a-f]{2})[:-]?([0-9a-f]{2})[:-]?([0-9a-f]{2})[:-]?([0-9a-f]{2})[:-]?([0-9a-f]{2})$')  # pylint: disable=line-too-long
_NETMASK_REGEX = re.compile(r'^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$')  # pylint: disable=line-too-long
_URL_REGEX = re.compile(r'^[a-z0-9]+:\/\/[/a-zA-Z0-9@:%._\+~#?&=\[\]]+')
_OPSI_HOST_KEY_REGEX = re.compile(r'^[0-9a-f]{32}$')
_PRODUCT_VERSION_REGEX = re.compile(r'^[a-z0-9.]{1,32}$')
_PACKAGE_VERSION_REGEX = re.compile(r'^[a-z0-9.]{1,16}$')
_PRODUCT_ID_REGEX = re.compile(r'^[a-z0-9-_\.]{1,128}$')
_PACKAGE_CUSTOM_NAME_REGEX = re.compile(r'^[a-zA-Z0-9]+$')
_PRODUCT_PROPERTY_ID_REGEX = re.compile(r'^\S+$')
_CONFIG_ID_REGEX = re.compile(r'^\S+$')
_GROUP_ID_REGEX = re.compile(r'^[a-z0-9][a-z0-9-_. ]*$')
_OBJECT_ID_REGEX = re.compile(r'^[a-z0-9][a-z0-9-_. ]*$')
_EMAIL_REGEX = re.compile(r'^(([A-Za-z0-9]+_+)|([A-Za-z0-9]+\-+)|([A-Za-z0-9]+\.+)|([A-Za-z0-9]+\++))*[A-Za-z0-9]+@((\w+\-+)|(\w+\.))*\w*')
_DOMAIN_REGEX = re.compile(r'^((\w+\-+)|(\w+\.))*\w{1,63}\.\w{2,16}\.?$')
_HOSTNAME_REGEX = re.compile(r'^[a-z0-9][a-z0-9\-]*$')
_LICENSE_CONTRACT_ID_REGEX = re.compile(r'^[a-z0-9][a-z0-9-_. :]*$')
_SOFTWARE_LICENSE_ID_REGEX = re.compile(r'^[a-z0-9][a-z0-9-_. :]*$')
_LICENSE_POOL_ID_REGEX = re.compile(r'^[a-z0-9][a-z0-9-_. :]*$')
_LANGUAGE_CODE_REGEX = re.compile(r'^([a-z]{2,3})[-_]?([a-z]{4})?[-_]?([a-z]{2})?$')
_ARCHITECTURE_REGEX = re.compile(r'^(x86|x64)$')


def forceList(var):
	if not isinstance(var, (set, list, tuple, types.GeneratorType)):
		return [var]

	return list(var)


def forceUnicode(var):  # pylint: disable=too-many-return-statements
	if isinstance(var, str):
		return var
	if (os.name == 'nt') and isinstance(var, WindowsError):  # pylint: disable=undefined-variable
		try:
			return f"[Error {var.args[0]}] {var.args[1]}"
		except Exception:  # pylint: disable=broad-except
			return str(var)
	try:
		if isinstance(var, bytes):
			return var.decode()
	except Exception:  # pylint: disable=broad-except
		pass

	try:
		var = var.__repr__()
		if isinstance(var, str):
			return var
		return str(var, 'utf-8', 'replace')
	except Exception:  # pylint: disable=broad-except
		pass

	return str(var)


def forceUnicodeLower(var):
	return forceUnicode(var).lower()


def forceUnicodeUpper(var):
	return forceUnicode(var).upper()


def forceUnicodeList(var):
	return [forceUnicode(element) for element in forceList(var)]


def forceDictList(var):
	return [forceDict(element) for element in forceList(var)]


def forceUnicodeLowerList(var):
	return [forceUnicodeLower(element) for element in forceList(var)]


def forceBool(var):
	if isinstance(var, bool):
		return var
	if isinstance(var, str):
		if len(var) <= 5:  # longest word is 5 characters ("false")
			lowValue = var.lower()
			if lowValue in ('true', 'yes', 'on', '1'):
				return True
			if lowValue in ('false', 'no', 'off', '0'):
				return False

	return bool(var)


def forceBoolList(var):
	return [forceBool(element) for element in forceList(var)]


def forceInt(var):
	if isinstance(var, int):
		return var
	try:
		return int(var)
	except Exception as err:
		raise ValueError(f"Bad int value '{var}': {err}") from err


def forceIntList(var):
	return [forceInt(element) for element in forceList(var)]


def forceUnsignedInt(var):
	var = forceInt(var)
	if var < 0:
		var = -1 * var
	return var


def forceOct(var):
	if isinstance(var, int):
		return var

	try:
		octValue = ''
		for idx, val in enumerate(forceUnicode(var)):
			val = forceInt(val)
			if val > 7:
				raise ValueError(f"{val} is too big")
			if idx == 0 and val != '0':
				octValue += '0'
			octValue += str(val)

		octValue = int(octValue, 8)
		return octValue
	except Exception as err:  # pylint: disable=broad-except
		raise ValueError(f"Bad oct value {var}: {err}") from err


def forceFloat(var):
	if isinstance(var, float):
		return var

	try:
		return float(var)
	except Exception as err:  # pylint: disable=broad-except
		raise ValueError(f"Bad float value '{var}': {err}") from err


def forceDict(var):
	if var is None:
		return {}
	if isinstance(var, dict):
		return var
	raise ValueError(f"Not a dict '{var}'")


def forceTime(var):
	"""
	Convert `var` to a time.struct_time.

	If no conversion is possible a `ValueError` will be raised.
	"""
	if isinstance(var, time.struct_time):
		return var
	if isinstance(var, datetime.datetime):
		var = time.mktime(var.timetuple()) + var.microsecond / 1E6

	if isinstance(var, (int, float)):
		return time.localtime(var)

	raise ValueError(f"Not a time {var}")


def forceHardwareVendorId(var):
	var = forceUnicodeUpper(var)
	if not re.search(_HARDWARE_ID_REGEX, var):
		raise ValueError(f"Bad hardware vendor id '{var}'")
	return var


def forceHardwareDeviceId(var):
	var = forceUnicodeUpper(var)
	if not re.search(_HARDWARE_ID_REGEX, var):
		raise ValueError(f"Bad hardware device id '{var}'")
	return var


def forceOpsiTimestamp(var):
	"""
	Make `var` an opsi-compatible timestamp.

	This is a string with the format "YYYY-MM-DD HH:MM:SS".

	If a conversion is not possible a `ValueError` will be raised.
	"""
	if not var:
		return '0000-00-00 00:00:00'
	if isinstance(var, datetime.datetime):
		return forceUnicode(var.strftime('%Y-%m-%d %H:%M:%S'))

	var = forceUnicode(var)
	match = re.search(_OPSI_TIMESTAMP_REGEX, var)
	if not match:
		match = re.search(_OPSI_DATE_REGEX, var)
		if not match:
			raise ValueError(f"Bad opsi timestamp: {var}")
		return f"{match.group(1)}-{match.group(2)}-{match.group(3)} 00:00:00"
	return (
		f"{match.group(1)}-{match.group(2)}-{match.group(3)}"
		f" {match.group(4)}:{match.group(5)}:{match.group(6)}"
	)


def forceFqdn(var):
	var = forceObjectId(var)
	if not _FQDN_REGEX.search(var):
		raise ValueError(f"Bad fqdn: '{var}'")
	if var.endswith('.'):
		var = var[:-1]
	return var

forceHostId = forceFqdn


def forceHostIdList(var):
	return [forceHostId(element) for element in forceList(var)]


def forceHardwareAddress(var):
	var = forceUnicodeLower(var)
	if not var:
		return var

	match = re.search(_HARDWARE_ADDRESS_REGEX, var)
	if not match:
		raise ValueError(f"Invalid hardware address: {var}")

	return (
		f"{match.group(1)}:{match.group(2)}:{match.group(3)}:"
		f"{match.group(4)}:{match.group(5)}:{match.group(6)}"
	).lower()

def forceIPAddress(var):
	if not isinstance(var, (ipaddress.IPv4Address, ipaddress.IPv6Address, str)):
		raise ValueError(f"Invalid ip address: '{var}'")
	var = ipaddress.ip_address(var)
	if isinstance(var, ipaddress.IPv6Address) and var.ipv4_mapped:
		return var.ipv4_mapped.compressed
	return var.compressed

forceIpAddress = forceIPAddress


def forceHostAddress(var):
	var = forceUnicodeLower(var)
	try:
		try:
			try:
				var = forceIpAddress(var)
			except Exception:  # pylint: disable=broad-except
				var = forceFqdn(var)
		except Exception:  # pylint: disable=broad-except
			var = forceHostname(var)
	except Exception as err:  # pylint: disable=broad-except
		raise ValueError(f"Invalid host address: '{var}'") from err
	return var


def forceNetmask(var):
	var = forceUnicodeLower(var)
	if not re.search(_NETMASK_REGEX, var):
		raise ValueError(f"Invalid netmask: '{var}'")
	return var


def forceNetworkAddress(var):
	if not isinstance(var, (ipaddress.IPv4Network, ipaddress.IPv6Network, str)):
		raise ValueError(f"Invalid network address: '{var}'")
	return ipaddress.ip_network(var).compressed


def forceUrl(var):
	"""
	Forces ``var`` to be an valid URL.

	:rtype: str
	"""
	var = forceUnicode(var)
	if not _URL_REGEX.search(var):
		raise ValueError(f"Bad url: '{var}'")
	return var


def forceOpsiHostKey(var):
	var = forceUnicodeLower(var)
	if not re.search(_OPSI_HOST_KEY_REGEX, var):
		raise ValueError(f"Bad opsi host key: {var}")
	return var


def forceProductVersion(var):
	var = forceUnicode(var)
	if not _PRODUCT_VERSION_REGEX.search(var):
		raise ValueError(f"Bad product version: '{var}'")
	return var


def forceProductVersionList(var):
	return [forceProductVersion(element) for element in forceList(var)]


def forcePackageVersion(var):
	var = forceUnicode(var)
	if not _PACKAGE_VERSION_REGEX.search(var):
		raise ValueError(f"Bad package version: '{var}'")
	return var


def forcePackageVersionList(var):
	return [forcePackageVersion(element) for element in forceList(var)]


def forceProductId(var):
	var = forceObjectId(var)
	if not _PRODUCT_ID_REGEX.search(var):
		raise ValueError(f"Bad product id: '{var}'")
	return var


def forceProductIdList(var):
	return [forceProductId(element) for element in forceList(var)]


def forcePackageCustomName(var):
	var = forceUnicodeLower(var)
	if not _PACKAGE_CUSTOM_NAME_REGEX.search(var):
		raise ValueError(f"Bad package custom name: '{var}'")
	return var


def forceProductType(var):
	lowercaseVar = forceUnicodeLower(var)
	if lowercaseVar in ('localboot', 'localbootproduct'):
		return 'LocalbootProduct'
	if lowercaseVar in ('netboot', 'netbootproduct'):
		return 'NetbootProduct'
	raise ValueError(f"Unknown product type: '{var}'")


def forceProductPropertyId(var):
	var = forceUnicodeLower(var)
	if not _PRODUCT_PROPERTY_ID_REGEX.search(var):
		raise ValueError(f"Bad product property id: '{var}'")
	return var


def forceConfigId(var):
	var = forceUnicodeLower(var)
	if not _CONFIG_ID_REGEX.search(var):
		raise ValueError(f"Bad config id: '{var}'")
	return var


def forceProductPropertyType(var):
	value = forceUnicodeLower(var)
	if value in ('unicode', 'unicodeproductproperty'):
		return 'UnicodeProductProperty'
	if value in ('bool', 'boolproductproperty'):
		return 'BoolProductProperty'
	raise ValueError(f"Unknown product property type: '{var}'")


def forceProductPriority(var):
	var = forceInt(var)
	if var < -100:
		return -100
	if var > 100:
		return 100
	return var


def forceFilename(var):
	return forceUnicode(var)


def forceProductTargetConfiguration(var):
	var = forceUnicodeLower(var)
	if var and var not in ('installed', 'always', 'forbidden', 'undefined'):
		raise ValueError(f"Bad product target configuration: '{var}'")
	return var


def forceInstallationStatus(var):
	var = forceUnicodeLower(var)
	if var and var not in ('installed', 'not_installed', 'unknown'):
		raise ValueError(f"Bad installation status: '{var}'")
	return var


def forceActionRequest(var):
	var = forceUnicodeLower(var)
	if var:
		if var == 'undefined':
			var = None
		elif var not in ('setup', 'uninstall', 'update', 'always', 'once', 'custom', 'none'):
			raise ValueError(f"Bad action request: '{var}'")
	return var


def forceActionRequestList(var):
	return [forceActionRequest(element) for element in forceList(var)]


def forceActionProgress(var):
	return forceUnicode(var)


def forceActionResult(var):
	var = forceUnicodeLower(var)
	if var and var not in ('failed', 'successful', 'none'):
		raise ValueError(f"Bad action result: '{var}'")
	return var


def forceRequirementType(var):
	var = forceUnicodeLower(var)
	if not var:
		return None
	if var not in ('before', 'after'):
		raise ValueError(f"Bad requirement type: '{var}'")
	return var


def forceObjectClass(var, objectClass):
	if isinstance(var, objectClass):
		return var

	exception = None
	if isinstance(var, str) and var.lstrip().startswith('{'):
		from OPSI.Util import fromJson  # pylint: disable=import-outside-toplevel
		try:
			var = fromJson(var)
		except Exception as err:  # pylint: disable=broad-except
			exception = err
			logger.debug("Failed to get object from json %s: %s", var, err)

	if isinstance(var, dict):
		if 'type' not in var:
			raise ValueError(f"Key 'type' missing in hash {var}")

		import OPSI.Object  # pylint: disable=import-outside-toplevel,unused-import
		try:
			_class = eval('OPSI.Object.%s' % var['type'])  # pylint: disable=eval-used
			if issubclass(_class, objectClass):
				var = _class.fromHash(var)
		except AttributeError as err:
			if "module 'OPSI.Object' has no attribute" in str(err):
				err = ValueError(f"Invalild object type: {var['type']}")

			exception = err
			logger.debug("Failed to get object from dict %s: %s", var, err)
		except Exception as err:  # pylint: disable=broad-except
			exception = err
			logger.debug("Failed to get object from dict %s: %s", var, err)

	if not isinstance(var, objectClass):
		if exception is None:
			raise ValueError(f"'{var}' is not a {objectClass}")
		raise ValueError(f"'{var}' is not a {objectClass}: {exception}")

	return var


def forceObjectClassList(var, objectClass):
	return [forceObjectClass(element, objectClass) for element in forceList(var)]


def forceGroupId(var):
	var = forceObjectId(var)
	if not _GROUP_ID_REGEX.search(var):
		raise ValueError(f"Bad group id: '{var}'")
	return var


def forceGroupType(var):
	lowercaseValue = forceUnicodeLower(var)

	if lowercaseValue == 'hostgroup':
		return 'HostGroup'
	if lowercaseValue == 'productgroup':
		return 'ProductGroup'
	raise ValueError(f"Unknown group type: '{var}'")


def forceGroupTypeList(var):
	return [forceGroupType(element) for element in forceList(var)]


def forceGroupIdList(var):
	return [forceGroupId(element) for element in forceList(var)]


def forceObjectId(var):
	var = forceUnicodeLower(var).strip()
	if not _OBJECT_ID_REGEX.search(var):
		raise ValueError(f"Bad object id: '{var}'")
	return var


def forceObjectIdList(var):
	return [forceObjectId(element) for element in forceList(var)]


def forceEmailAddress(var):
	var = forceUnicodeLower(var)
	if not _EMAIL_REGEX.search(var):
		raise ValueError(f"Bad email address: '{var}'")
	return var


def forceDomain(var):
	var = forceUnicodeLower(var)
	if not _DOMAIN_REGEX.search(var):
		raise ValueError(f"Bad domain: '{var}'")
	return var


def forceHostname(var):
	var = forceUnicodeLower(var)
	if not _HOSTNAME_REGEX.search(var):
		raise ValueError(f"Bad hostname: '{var}'")
	return var


def forceLicenseContractId(var):
	var = forceUnicodeLower(var)
	if not _LICENSE_CONTRACT_ID_REGEX.search(var):
		raise ValueError(f"Bad license contract id: '{var}'")
	return var


def forceLicenseContractIdList(var):
	return [forceLicenseContractId(element) for element in forceList(var)]


def forceSoftwareLicenseId(var):
	var = forceUnicodeLower(var)
	if not _SOFTWARE_LICENSE_ID_REGEX.search(var):
		raise ValueError(f"Bad software license id: '{var}'")
	return var


def forceSoftwareLicenseIdList(var):
	return [forceSoftwareLicenseId(element) for element in forceList(var)]


def forceLicensePoolId(var):
	var = forceUnicodeLower(var)
	if not _LICENSE_POOL_ID_REGEX.search(var):
		raise ValueError(f"Bad license pool id: '{var}'")
	return var


def forceLicensePoolIdList(var):
	return [forceLicensePoolId(element) for element in forceList(var)]


def forceAuditState(var):
	var = forceInt(var)
	if var not in (0, 1):
		raise ValueError(f"Bad audit state value: {var}")
	return var


def forceLanguageCode(var):
	var = forceUnicodeLower(var)
	match = _LANGUAGE_CODE_REGEX.search(var)
	if not match:
		raise ValueError(f"Bad language code: '{var}'")
	var = match.group(1)
	if match.group(2):
		var = f"{var}-{match.group(2).capitalize()}"
	if match.group(3):
		var = f"{var}-{match.group(3).upper()}"
	return var


def forceLanguageCodeList(var):
	return [forceLanguageCode(element) for element in forceList(var)]


def forceArchitecture(var):
	var = forceUnicodeLower(var)
	if not _ARCHITECTURE_REGEX.search(var):
		raise ValueError(f"Bad architecture: '{var}'")
	return var


def forceArchitectureList(var):
	return [forceArchitecture(element) for element in forceList(var)]


def forceUniqueList(_list):
	cleanedList = []
	for entry in _list:
		if entry not in cleanedList:
			cleanedList.append(entry)
	return cleanedList


def args(*vars, **typeVars):  # pylint: disable=redefined-builtin
	"""Function to populate an object with passed on keyword args.
	This is intended to be used as a decorator.
	Classes using this decorator must explicitly inherit from object or a subclass of object.

	.. code-block:: python

		@args()			#works
		class Foo(object):
			pass

		@args()			#works
		class Bar(Foo):
			pass

		@args()			#does not work
		class Foo():
			pass

		@args()			#does not work
		class Foo:
			pass
	"""
	vars = list(vars)

	def wrapper(cls):
		def new(typ, *args, **kwargs):  # pylint: disable=redefined-builtin,redefined-outer-name
			if getattr(cls, "__base__", None) in (object, None):
				obj = object.__new__(typ)  # Suppress deprecation warning
			else:
				obj = cls.__base__.__new__(typ, *args, **kwargs)

			vars.extend(list(typeVars.keys()))
			ka = kwargs.copy()

			for var in vars:
				varName = var.lstrip("_")
				if varName in ka:
					if var in typeVars:
						func = typeVars[var]
						ka[var] = func(ka[varName])
					else:
						ka[var] = ka[varName]
				else:
					ka[var] = None

			for key, value in ka.items():
				if getattr(obj, key, None) is None:
					setattr(obj, key, value)

			return obj

		cls.__new__ = staticmethod(new)
		return cls

	return wrapper
